### PR TITLE
Package LICENSE with each component

### LINKED ISSUE #771: LICENSE file not packaged with individual crates
In our Cargo.toml files we have a `license-file` with a relative path back to the root LICENSE file. This means that LICENSE is not actually packaged with the individual crates, which breaks tooling that expects to be able read the LICENSE file, like the `vendor_rust.py` script used to vendor crates into Firefox.

Assuming we can't use `license` in Cargo.toml to `OR` together the different licenses we're using, I think that we'll have to include a copy of the LICENSE file with each component that we package as a separate crate.

### PR DESCRIPTION
**Pull request by dminor**
This packages LICENSE with each component and adds commands to automatically copy LICENSE files to each component and a check to make sure LICENSE exists in each component.

Fixes #771.

### CODE REVIEW

File: tools/scripts/tidy.toml [hunk 1]
```diff
@@ -98,4 +98,46 @@ if greater_than ${output_length} 0
...
+exit_on_error true
+
+glob_pattern = set "./**/Cargo.toml"
+license = canonicalize LICENSE
+
+handle = glob_array ${glob_pattern}
+for path in ${handle}
+    root_dir = dirname ${path}
+    if not is_empty ${root_dir}
+        if not is_path_exists ${root_dir}/LICENSE
```
Comment (Manishearth) [inline tools/scripts/tidy.toml:137]: suggestion (nb): should we check that it is the same license, too?

File: tools/scripts/tidy.toml [hunk 2]
```diff
@@ -98,4 +98,46 @@ if greater_than ${output_length} 0
...
+for path in ${handle}
+    root_dir = dirname ${path}
+    if not is_empty ${root_dir}
+        echo "Copying LICENSE to ${root_dir}/LICENSE"
+        cp ${license} ${root_dir}/LICENSE
+    end
+end
+'''
+
+[tasks.license-file-check]
```
Comment (Manishearth) [inline tools/scripts/tidy.toml:123]: This should be a part of the ci-task-tidy task in the toplevel Makefile

File: tools/scripts/tidy.toml [hunk 3]
```diff
@@ -98,4 +98,46 @@ if greater_than ${output_length} 0
...
+
+glob_pattern = set "./**/Cargo.toml"
+license = canonicalize LICENSE
+
+handle = glob_array ${glob_pattern}
+for path in ${handle}
+    root_dir = dirname ${path}
+    if not is_empty ${root_dir}
+        if not is_path_exists ${root_dir}/LICENSE
+            trigger_error "LICENSE file missing in ${root_dir}"
```
Comment (Manishearth) [inline tools/scripts/tidy.toml:138]: suggestion: mention something like "run `cargo make copy-license` to automatically copy the license to all necessary locations"

**Review by Manishearth — APPROVED** [review]

File: tools/scripts/tidy.toml (hunk 1 above)
Comment (srl295) [reply tools/scripts/tidy.toml:137]: something like `if cmp -s ${root_dir}/LICENSE ${root_dir}/../LICENSE` (don't know my duckscript)

File: tools/scripts/tidy.toml (hunk 3 above)
Comment (srl295) [reply tools/scripts/tidy.toml:138]: is there any linter on the lower level .toml files themselves? maybe something to enforce consistency (the all have the same `categories` etc) would be useful.

**Review by srl295 — APPROVED** [review]

File: tools/scripts/tidy.toml (hunk 3 above)
Comment (Manishearth) [reply tools/scripts/tidy.toml:138]: Not yet, but that's a good idea!

**Comment by zbraniecki** [comment]
Non blocking: I really wish cargo would know how to package in files from `../../` - this seems important for mono repo licenses and the solution here is really a dirty hack that increases the cost on the project to work around cargo's limitation.

@Manishearth do you agree? Could we file an issue against cargo?

**Review by zbraniecki — DISMISSED** [review]

File: tools/scripts/tidy.toml (hunk 1 above)
Comment (dminor) [reply tools/scripts/tidy.toml:137]: I tried to do this prior to pushing this up for the initial review, but I was unable to get it working in duckscript:
```
license_text = readfile ${license}

if not eq ${license_text} ${license_text}
    echo "not working"
end
```
but it turns out that this works:
```
license_text = readfile ${license}

license_equal = eq ${license_text} ${license_text}
if not eq ${license_equal}
    echo "not working"
end
```
Duckscript is not intuitive to me at all.

**Comment by Manishearth** [comment]
Yes, please do! I'm not sure how it should work but worth filing an issue.

**Review by Manishearth — APPROVED** [review]

**Comment by zbraniecki** [comment]
Filed as https://github.com/rust-lang/cargo/issues/9575

**Comment by dminor** [comment]
It looks like this is actually a problem when running `cargo vendor`, not with the packaging. I wish I had noticed that earlier!

**Comment by zbraniecki** [comment]
Agree.

**Comment by Manishearth** [comment]
@dminor That's not clear to me at all; it seems like a minor bug in cargo vendor to me so far. I'd rather avoid doing this if we don't have to.

**Comment by dminor** [comment]
> @dminor That's not clear to me at all; it seems like a minor bug in cargo vendor to me so far. I'd rather avoid doing this if we don't have to.

I agree, I'd prefer to not do this either. Are you willing to fix the upstream bug? If so, I'm happy to close this again and move on with my life :)

**Comment by Manishearth** [comment]
I don't have the time or the knowledge to fix the upstream bug; but I'm pushing on it.

**Comment by dminor** [comment]
From conversation on slack, the plan is to go ahead with this workaround until the upstream cargo issue is fixed.

**Review by Manishearth — APPROVED** [review]

**Review by sffc — APPROVED** [review]

**Review by zbraniecki — APPROVED** [review]

**Review by gregtatum — APPROVED** [review]

**Pull request merged**